### PR TITLE
feat: allow setting endpoint url through env and shared config

### LINKED ISSUE #1019: Service specific endpoint URL config support
### Describe the feature

Support environment variable and shared config settings for service specific URLs: https://docs.aws.amazon.com/sdkref/latest/guide/feature-ss-endpoints.html

### Is your Feature Request related to a problem?

See https://docs.aws.amazon.com/sdkref/latest/guide/feature-ss-endpoints.html

### Proposed Solution

See the SEP

### Describe alternative solutions or features you've considered

_No response_

### Acknowledge

- [X] I may be able to implement this feature request

### AWS Kotlin SDK version used

n/a

### Platform (JVM/JS/Native)

n/a

### Operating System and version

n/a

### PR DESCRIPTION
**Pull request by lucix-aws**
## Issue \#
Closes #1019 

## Description of changes
Implements endpoint URL configuration through env/shared config.

Endpoint URLs are configured at either the global or service level and are respected in the following order:

* directly configured in code
* service-specific environment setting (sysprop, envvar)
* global environment setting
* service-specific config setting
* global config setting

This resolution can also be disabled globally through an environment or config setting.

misc:

* Refactor `ServiceClientCompanionObjectWriter` to have the base behavior of generating config-based endpoint URL resolution.
* `AbstractAwsSdkClientFactory::finalizeConfig` now takes an entire (lazy) shared config instead of just the active profile.
* Added (and consumed) explicit implementations of the standardized sdkId transforms from the serviceId step in codegen.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/SdkIdTransform.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * Implements a single standardized sdkId transform.
+ */
+interface SdkIdTransformer {
+    fun transform(id: String): String
+}
+
+/**
+ * Implements all standardized sdkId transforms.
+ */
+object SdkIdTransform {
```
Comment (aajtodd) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/SdkIdTransform.kt:17]: **nit**: It looks like only `LowerSnakeCase` and `UpperSnakeCase` are the only transforms used

**Review by aajtodd — APPROVED** [review]
Nice work.

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/SdkIdTransform.kt (hunk 1 above)
Comment (lucix-aws) [reply codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/SdkIdTransform.kt:17]: Yeah, the remainders were more for completeness (the SEP lists these exact 5). I'm fine with removing them though.

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/SdkIdTransform.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        override fun transform(id: String): String = id.replaceWhitespace("_").uppercase()
+    }
+}
+
+/**
+ * Applies a concrete sdkId transform to a string.
+ */
+fun String.withTransform(transformer: SdkIdTransformer): String = transformer.transform(this)
+
+private fun String.replaceWhitespace(replacement: String) = replace(Regex("\\s"), replacement)
```
Comment (ianbotsf) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/SdkIdTransform.kt:59]: **Nit**: `Regex("\\s")` should be a top-level constant.

**Review by ianbotsf — APPROVED** [review]
> Endpoint URLs are configured at either the global or service level and are respected in the following order:
> 
> * directly configured in code
> * service-specific environment setting (sysprop, envvar)
> * global environment setting
> * service-specific config setting
> * global config setting
>
> This resolution can also be disabled globally through an environment or config setting.

~~This sounds complicated. Is this the kind of thing we should document in our dev guide rather than just in the KDocs of an `@InternalSdk` method?~~

EDIT: I have been informed that [this _is_ already documented in the dev guide](https://docs.aws.amazon.com/sdkref/latest/guide/feature-ss-endpoints.html#ss-endpoints-precedence). My apologies.

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/profile/ServicesConfig.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.sdk.kotlin.runtime.config.profile
+
+import aws.sdk.kotlin.runtime.InternalSdkApi
+
+@InternalSdkApi
+public typealias ServicesConfig = ConfigSection
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/profile/ServicesConfig.kt:11]: This was to follow what we have for all of our sections now - note that `SsoSectionConfig` is defined in much the same way (it has a standalone type alias, which has no extensions, but an `AwsSharedConfig` has a getter for matching sections where the type is used in the return signature).

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/AwsSdkSetting.kt [hunk 4]
```diff
@@ -152,4 +154,28 @@ public object AwsSdkSetting {
...
+ * Resolves an endpoint url for a given service.
+ */
+public fun AwsSdkSetting.resolveEndpointUrl(
+    provider: PlatformProvider,
+    sysPropSuffix: String,
+    envSuffix: String,
+): Url? {
+    val serviceSetting = EnvironmentSetting(Url::parse)("aws.endpointUrl$sysPropSuffix", "AWS_ENDPOINT_URL_$envSuffix")
+    return serviceSetting.resolve(provider) ?: AwsEndpointUrl.resolve(provider)
 }
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/AwsSdkSetting.kt:181]: `@InternalSdkApi` was a miss though I imagine `AwsSdkSetting` _itself_ being marked internal would cover it. I've added it regardless.

As for where the logic lives, this is just how I chose to shake things out organizationally (in both this and `AwsSharedConfig` - they both have a self-contained API to resolve endpoint URL). Having `resolveEndpointUrl` delegate directly to all 4 variants would been fine, it just made sense to me to compartmentalize each of those at the specific configs. I don't have strong feelings about that breakdown.

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/profile/ServicesConfig.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ianbotsf) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/profile/ServicesConfig.kt:11]: Note that [type alias uniqueness is not compile-time enforced](https://kotlinlang.org/docs/type-aliases.html#:~:text=Type%20aliases%20do,and%20vice%20versa%3A). Returning or accepting a `ServicesConfig` or `SsoSessionConfig` type alias does nothing to enforce that a different alias isn't assigned/passed.

 The _only_ benefit type aliases have is to shorten the names of long types. In this case, both `ServicesConfig` and `SsoSessionConfig` are longer than `ConfigSection`.

**Pull request merged**